src/drivers/nvme.c

```diff
@@ -0,0 +1,865 @@
+        list_delete(l);
+        spin_unlock_irq(&n->lock, irqflags);
+        nvme_ioreq req = struct_from_list(l, nvme_ioreq, l);
+        apply(req->sh, (req->sc == NVME_SC_OK) ? STATUS_OK :
+                timm("result", "NVMe status code 0x%x", req->sc));
+        irqflags = spin_lock_irq(&n->lock);
+        list_insert_before(list_begin(&n->free_reqs), l);
+    }
+    nvme_service_pending(n, true);
+    spin_unlock_irq(&n->lock, irqflags);
+}
+
+closure_function(4, 0, void, nvme_ns_attach,
+                 nvme, n, u32, ns_id, u64, disk_size, storage_attach, a)
+{
+    nvme n = bound(n);
+    u32 ns_id = bound(ns_id);
+    u64 disk_size = bound(disk_size);
+    block_io r = closure(n->general, nvme_io, n, ns_id, false);
+    if (r == INVALID_ADDRESS) {
+        msg_err("failed to allocate read closure\n");
+        goto done;
+    }
+    block_io w = closure(n->general, nvme_io, n, ns_id, true);
+    if (w != INVALID_ADDRESS) {
+        nvme_debug("attaching disk (NS ID %d, capacity %ld bytes)", ns_id, disk_size);
+        apply(bound(a), r, w, disk_size);
+    } else {
+        msg_err("failed to allocate write closure\n");
+        deallocate_closure(r);
+    }
+  done:
+    closure_finish();
+}
+
+static void nvme_ns_resp_parse(nvme n, u32 ns_id, void *ns_resp,
+                               struct nvme_cqe *cqe, storage_attach a)
+{
+    nvme_debug("identify namespace response (cmd ID %d)", NVME_CMD_ID(cqe->dw3));
+    n->asq.head = NVME_SQ_HEAD(cqe->dw2);
+    int sc = NVME_STATUS_CODE(cqe->dw3);
+    if (sc != NVME_SC_OK) {
+        msg_err("failed to identify namespace: status code 0x%x\n", sc);
+        for (int i = 0; i < U64_FROM_BIT(n->asq.order); i++) {
+            struct nvme_sqe *sqe = &n->asq.ring[i];
+            rprintf("sqe %x %x %x %x %x %x %x %x %x %x %x %x\n",
+                    sqe->cdw0, sqe->nsid, sqe->reserved, sqe->mptr, sqe->dptr.prp1, sqe->dptr.prp2,
+                    sqe->cdw10, sqe->cdw11, sqe->cdw12, sqe->cdw13, sqe->cdw14, sqe->cdw15);
+        }
+        for (int i = 0; i < U64_FROM_BIT(n->acq.order); i++) {
+            struct nvme_cqe *cqe = &n->acq.ring[i];
+            rprintf("cqe %x %x %x %x\n", cqe->dw0, cqe->dw1, cqe->dw2, cqe->dw3);
+        }
+        return;
+    }
+    int flbas = *(u8 *)(ns_resp + 26) & 0xF;    /* formatted LBA size */
+    u32 lbaf = *(u32 *)(ns_resp + 128 + flbas); /* LBA format */
+    int lbads = 1 << ((lbaf >> 16) & 0xFF);     /* LBA data size */
+    if (lbads != SECTOR_SIZE) {
+        msg_err("unsupported sector size %ld", lbads);
+        return;
+    }
+    u64 disk_size = *(u64 *)ns_resp * lbads;
+    thunk ns_attach = closure(n->general, nvme_ns_attach, n, ns_id, disk_size, a);
+    if (ns_attach == INVALID_ADDRESS)
+        msg_err("failed to allocate NS attach closure\n");
+    else
+        enqueue(runqueue, ns_attach);
+}
+
+static void nvme_ns_query(nvme n, u32 ns_id, void *ns_resp)
+{
+    struct nvme_sqe *cmd = nvme_get_sqe(&n->asq);
```

Comment:
Could return 0? (or assert non-zero)

Reply:
Commands are sent to the admin queue one at a time, and each command is sent after receiving the response to the previous command, so there should always be an available submission queue entry. I added an `assert(cmd)` here and in similar places.